### PR TITLE
[Community] reference only control

### PR DESCRIPTION
**Pull request by okotaku**
Refer to https://github.com/Mikubill/sd-webui-controlnet/discussions/1236

Reference Image

reference_image

Output Image of `reference_attn=True` and `reference_adain=False`

output_image

Output Image of `reference_attn=False` and `reference_adain=True`

output_image

Output Image of `reference_attn=True` and `reference_adain=True`

output_image

### CODE REVIEW

**Comment by jfischoff** [comment]
Should we include the group norm modification in this PR as well: https://github.com/Mikubill/sd-webui-controlnet/pull/1278/files#diff-8c8d004eed5a3078434f6fbde15c178e472565ebfcb3119f308f9292c8eb7514R458 ?

Right now using the group norm or reference_adain+attn gives the best results for reference only. I would definitely like to see this get added either in this commit or a subsequent one.

**Review by patrickvonplaten — APPROVED** [review]
Works for me! Thanks!

**Comment by kadirnar** [comment]
Will you add controlnet support? How can we use it with controlnet? @okotaku

**Comment by wangdong-ivymobile** [comment]
It doesn't work for me for `EulerAncestralDiscreteScheduler`

**Comment by Laidawang** [comment]
When i run the example i get the error：
AttributeError: 'StableDiffusionReferencePipeline' object has no attribute 'image_processor'

**Comment by wangdong-ivymobile** [comment]
> When i run the example i get the error： AttributeError: 'StableDiffusionReferencePipeline' object has no attribute 'image_processor'

You can't use the version from PyPi, you need to pull master branch and install from local (that or wait for next release).

**Comment by Laidawang** [comment]
> > When i run the example i get the error： AttributeError: 'StableDiffusionReferencePipeline' object has no attribute 'image_processor'
> 
> You can't use the version from PyPi, you need to pull master branch and install from local (that or wait for next release).

thanks, i see

**Comment by okotaku** [comment]
@wangdong-ivymobile Thank you for your report. I fixed this bug on latest commit.

**Comment by jfischoff** [comment]
Just fyi, @lllyasviel fixed a bug recently: https://github.com/Mikubill/sd-webui-controlnet/pull/1309

**Comment by kadirnar** [comment]
> reference_adain

This is great 💯 Can we add inpaint feature like in this repo?
image
https://github.com/Mikubill/sd-webui-controlnet

**Comment by okotaku** [comment]
@jfischoff Thank you for your suggestion. I fixed style fidelity rule on latest commit, it is based on https://github.com/Mikubill/sd-webui-controlnet/pull/1309 .

**Comment by wangdong-ivymobile** [comment]
I tried to use this with multiple images (with slight modification), the result is very bad. I wonder if there are some tricks to make it better?
What I did is basically instead of using a latent of a single image, I concat all the ref image latent together. 
What I wish to achieve is maybe to have different ref images contributing differently to the result image, e.g. the first setting up the tone and theme, the second add some style and details, etc.

**Comment by jfischoff** [comment]
I'm getting slightly different results with this pipeline vs the automatic1111. 
This pipeline
result
Automatic1111
00003-1-A colorful Rococo painting of two female Tom Cruises dancing, with white dresses  Painting by Jean-Antoine Watteau

Looking at the code I notice the implementation is different between the original and this PR. I'm not sure it is responsible for the differences, but I was curious for instance why some of the changes were made. 

For instance the original code using the style blend to control adding sequences to the bank

```python
                if outer.attention_auto_machine == AutoMachine.Write:
                    if outer.attention_auto_machine_weight > self.attn_weight:
                        self.bank.append(self_attention_context.detach().clone())
                        self.style_cfgs.append(outer.current_style_fidelity)
                if outer.attention_auto_machine == AutoMachine.Read:
                    if len(self.bank) > 0:
                        style_cfg = sum(self.style_cfgs) / float(len(self.style_cfgs))
                        self_attn1_uc = self.attn1(x_norm1, context=torch.cat([self_attention_context] + self.bank, dim=1))
                        self_attn1_c = self_attn1_uc.clone()
                        if len(outer.current_uc_indices) > 0 and style_cfg > 1e-5:
                            self_attn1_c[outer.current_uc_indices] = self.attn1(
                                x_norm1[outer.current_uc_indices],
                                context=self_attention_context[outer.current_uc_indices])
                        self_attn1 = style_cfg * self_attn1_c + (1.0 - style_cfg) * self_attn1_uc
                    self.bank = []
                    self.style_cfgs = []
                if self_attn1 is None:
                    self_attn1 = self.attn1(x_norm1, **context=self_attention_context)**
```

Will this PR does the check during the `read` mode: 

```
if MODE == "write":
                    self.bank.append(norm_hidden_states.detach().clone())
                    attn_output = self.attn1(
                        norm_hidden_states,
                        encoder_hidden_states=encoder_hidden_states if self.only_cross_attention else None,
                        attention_mask=attention_mask,
                        **cross_attention_kwargs,
                    )
                if MODE == "read":
                    if attention_auto_machine_weight > self.attn_weight:
                        attn_output_uc = self.attn1(
                            norm_hidden_states,
                            encoder_hidden_states=torch.cat([norm_hidden_states] + self.bank, dim=1),
                            # attention_mask=attention_mask,
                            **cross_attention_kwargs,
                        )
                        attn_output_c = attn_output_uc.clone()
                        if do_classifier_free_guidance and style_fidelity > 0:
                            attn_output_c[uc_mask] = self.attn1(
                                norm_hidden_states[uc_mask],
                                encoder_hidden_states=norm_hidden_states[uc_mask],
                                **cross_attention_kwargs,
                            )
                        attn_output = style_fidelity * attn_output_c + (1.0 - style_fidelity) * attn_output_uc
                        self.bank.clear()
                    else:
                        attn_output = self.attn1(
                            norm_hidden_states,
                            encoder_hidden_states=encoder_hidden_states if self.only_cross_attention else None,
                            attention_mask=attention_mask,
                            **cross_attention_kwargs,
                        **)
```

I haven't looked closely enough to see if there are other differences, nor do I know if this is why my images are different. I'm just curious mostly.

**Comment by jfischoff** [comment]
FWIW I switched the `if outer.attention_auto_machine_weight > self.attn_weight` and it didn't change my image

**Comment by okotaku** [comment]
@jfischoff @wangdong-ivymobile 

The difference in code arises from the variations in the implementation of Unet and ControlNet between "diffusers" and "webui."

For example, if you look at the implementation of BasicTransformerBlock, you will notice minor differences in the code.
https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/ldm/modules/attention.py#L211
https://github.com/huggingface/diffusers/blob/85eff637aad1106f593d7535ec41cdb736b0b2ea/src/diffusers/models/attention.py#L121

Consequently, I made modifications to the implementation in "webui" to align it with the implementation in "diffusers." In terms of functionality, both implementations should achieve nearly the same results. One small difference to note is the implementation of "gn_weight" in this section, which varies slightly.
https://github.com/Mikubill/sd-webui-controlnet/blob/main/scripts/hook.py#L523-L536

This is due to the difference in the unit of blocks between "webui" and "diffusers." As a result, even with the same "gn_auto_machine_weight," the range of adain adaptation may differ slightly. However, I think this will not have a significant impact on the results, so I have simplified the implementation to match that of "diffusers."

File: examples/community/stable_diffusion_reference.py [hunk 1]
```diff
@@ -0,0 +1,772 @@
...
+        # 10. Denoising loop
+        num_warmup_steps = len(timesteps) - num_inference_steps * self.scheduler.order
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            for i, t in enumerate(timesteps):
+                # expand the latents if we are doing classifier free guidance
+                latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
+                latent_model_input = self.scheduler.scale_model_input(latent_model_input, t)
+
+                # ref only part
+                noise = torch.randn_like(ref_image_latents)
```
Comment (skirsten) [inline examples/community/stable_diffusion_reference.py:708]: I believe this should be 
```py
noise = randn_tensor(ref_image_latents.shape, generator=generator, device=ref_image_latents.device, dtype=ref_image_latents.dtype)
```
from the utils import to ensure the generation is deterministic.

**Comment by jfischoff** [comment]
@okotaku that makes sense. Thanks for answer my question.

**Comment by patrickvonplaten** [comment]
Cool let's merge this one!

**Comment by learningyan** [comment]
hi. TypeError: Transformer2DModel.forward() got an unexpected keyword argument 'attention_mask'  
also got an unexpected keyword argument 'encoder_attention_mask'  
How to solve it ? @okotaku

**Comment by blx0102** [comment]
By running the example case, got OUT OF MEMORY error, I used 3090 with 24G memory

**Comment by okotaku** [comment]
@learningyan https://github.com/huggingface/diffusers/pull/3508
I fixed the bug in this PR.
Please pull latest main branch.

**Comment by learningyan** [comment]
> 

i have pulled latest main branch, still have this question in line 501 and 502 of examples/community/stable_diffusion_controlnet_reference.py 

when I comment out these codes, the code is right. but the quality of  generated images is poor. So I'm not sure if commenting out these codes is correct. 

reference image :

input3

inference code:

pipe = StableDiffusionReferencePipeline.from_pretrained(
                "runwayml/stable-diffusion-v1-5",
                safety_checker=None,
                torch_dtype=torch.float16
                ).to('cuda:0')

pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)

result_img = pipe(ref_image=input_image,
                        prompt="1girl, masterpiece, best quality",
                        negative_prompt="lowres, bad anatomy, bad hands, text, error, missing fingers, extra digit, fewer digits, cropped, worst quality, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry",
                        num_inference_steps=20,
                        num_images_per_prompt=1,
                        reference_attn=True,
                        reference_adain=True).images[0]
result_img.save('reference.png')

result_img:

reference

 the results in sd-webui-controlnet (see https://github.com/Mikubill/sd-webui-controlnet/discussions/1236)

image

So How can I get the similar result as  sd-webui-controlnet?
 thanks a lot.

**Comment by okotaku** [comment]
@learningyan 

> i have pulled latest main branch, still have this question in line 501 and 502 of examples/community/stable_diffusion_controlnet_reference.py

You need to update diffusers too.

```
git pull
pip install .
```

> the quality of generated images is poor. 

You can change base model.
For example,

```
pipe = StableDiffusionReferencePipeline.from_pretrained(
"andite/anything-v4.0",
safety_checker=None,
torch_dtype=torch.float16
).to('cuda:0')
```

And this result is reference attention only,

```
result_img = pipe(ref_image=input_image,
prompt="1girl, masterpiece, best quality",
negative_prompt="lowres, bad anatomy, bad hands, text, error, missing fingers, extra digit, fewer digits, cropped, worst quality, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry",
num_inference_steps=20,
num_images_per_prompt=1,
reference_attn=True,
reference_adain=False,
style_fidelity=0.5  # you can set style_fidelity=1.0
).images[0]
```

https://github.com/Mikubill/sd-webui-controlnet/discussions/1236#discussioncomment-5891377

When examining the details more closely, it appears that the actual model being used is anythingv3 and a custom model named animevae.pt, etc.

**Comment by vijishmadhavan** [comment]
why I am getting kind of white tint on images using controlnet reference only?
image (12)
image (9)

`
pipe = StableDiffusionControlNetReferencePipeline.from_pretrained(
                "stabilityai/stable-diffusion-2-1",
                controlnet=controlnet,
                safety_checker=None,
                torch_dtype=torch.float16
                ).to('cuda:0')

pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)`
`    image1 = Image.fromarray(image)
    image = np.array(image)
    low_threshold = 100
    high_threshold = 200
    canny = cv2.Canny(image, low_threshold, high_threshold)
    canny = canny[:, :, None]
    canny = np.concatenate([canny, canny, canny], axis=2)
    canny_image = Image.fromarray(canny)
    result_img = pipe(ref_image=image1,
                            prompt=prompt,
                            negative_prompt="lowres, two head, bad anatomy, bad hands, text, error, missing fingers, extra digit, fewer digits, cropped, worst quality, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry",
                            image=canny_image,
                            num_inference_steps=20,
                            controlnet_conditioning_scale=0.8,
                            reference_attn=True,
                            reference_adain=True,
                            style_fidelity=0.5).images[0]`

**Comment by miiiz** [comment]
i just git from the lastest diffusers and install following the important note but got this error. StableDiffusionReferencePipeline works fine. 

AttributeError: 'StableDiffusionControlNetReferencePipeline' object has no attribute '_default_height_width'

**Comment by blx0102** [comment]
@miiiz you can copy the code of that function from reference file to the controlnet reference file

**Comment by okotaku** [comment]
@miiiz @blx0102 Thank you for your feedback.
I updated the codes.

https://github.com/huggingface/diffusers/pull/3723

**Comment by alexisrolland** [comment]
Hi there, this thread is a bit long and confusing. Is there any documentation  somewhere that describes the classes StableDiffusionReferencePipeline and StableDiffusionCobtrolNetReferencePipeline?

**Comment by gasvn** [comment]
@okotaku Thanks for your diffusers implementation. Based on your code, I achieve a cross-image region drag based on the reference scheme. 1. Use inpaint controlnet to extract inpainted region feature from another image. 2. Use the segment-anything controlnet to keep reasonable pose.

Code in https://github.com/sail-sg/EditAnything
image

**Comment by okotaku** [comment]
@gasvn Really cool projects! Thank you for your contribution.

**Comment by ethank852** [comment]
two questions here:
1. I see `StableDiffusionControlNetReferencePipeline` and `StableDiffusionReferencePipeline`. What are the differences?
2. I'm unable to import `StableDiffusionControlNetReferencePipeline` directly from the `diffusers` module. From [the doc](https://huggingface.co/docs/diffusers/using-diffusers/custom_pipeline_examples) here, looks like I should use `custom_pipeline` kwarg?

**Comment by patrickvonplaten** [comment]
Yes you need to use the `custom_pipeline` kwarg :-)

**Comment by djj0s3** [comment]
> two questions here:
> 
> 1. I see `StableDiffusionControlNetReferencePipeline` and `StableDiffusionReferencePipeline`. What are the differences?
> 2. I'm unable to import `StableDiffusionControlNetReferencePipeline` directly from the `diffusers` module. From [the doc](https://huggingface.co/docs/diffusers/using-diffusers/custom_pipeline_examples) here, looks like I should use `custom_pipeline` kwarg?

For anyone coming to this later, you'll definitely need custom_pipeline like so:
```
controlnet_reference_pipe = StableDiffusionPipeline.from_pretrained(
  MODEL_ID,
  cache_dir=MODEL_CACHE_DIR,
  custom_pipeline="stable_diffusion_controlnet_reference",
  controlnet=CONTROLNET_PIPE,
  safety_checker=None,
  torch_dtype=torch.float16
  )
```

**Comment by amrakm** [comment]
@djj0s3 does that mean you can't use reference-only with other ControlNets using `StableDiffusionMultiControlNetPipeline`

**Comment by djj0s3** [comment]
@amrakm are you trying to use reference-only as a controlnet to pass into another pipe with other controlnets? Interesting idea! One option you can try is to import this class directly and use it as `StableDiffusionControlNetReferencePipeline.from_pretrained(...):...`Since it is written as a subclass of StableDiffusionControlNetPipeline, you should be able to accomplish what you're trying to do. I'm coincidentally loading it this way vs using `custom_pipeline` since I had to tweak the class to get a couple of other things I'm doing working with it.  I haven't tried what you're doing, but that approach should work for you.

**Comment by Akmpfen** [comment]
I see the code has reference_attn and reference_adain, but how to use reference_only?

**Comment by dotieuthien** [comment]
> why I am getting kind of white tint on images using controlnet reference only? image (12) image (9)
> 
> ` pipe = StableDiffusionControlNetReferencePipeline.from_pretrained( "stabilityai/stable-diffusion-2-1", controlnet=controlnet, safety_checker=None, torch_dtype=torch.float16 ).to('cuda:0')
> 
> pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)` ` image1 = Image.fromarray(image) image = np.array(image) low_threshold = 100 high_threshold = 200 canny = cv2.Canny(image, low_threshold, high_threshold) canny = canny[:, :, None] canny = np.concatenate([canny, canny, canny], axis=2) canny_image = Image.fromarray(canny) result_img = pipe(ref_image=image1, prompt=prompt, negative_prompt="lowres, two head, bad anatomy, bad hands, text, error, missing fingers, extra digit, fewer digits, cropped, worst quality, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry", image=canny_image, num_inference_steps=20, controlnet_conditioning_scale=0.8, reference_attn=True, reference_adain=True, style_fidelity=0.5).images[0]`
I face with the same issue. Could you please give me any advices? @okotaku

**Comment by chyohoo** [comment]
> > why I am getting kind of white tint on images using controlnet reference only? image (12) image (9)
> > ` pipe = StableDiffusionControlNetReferencePipeline.from_pretrained( "stabilityai/stable-diffusion-2-1", controlnet=controlnet, safety_checker=None, torch_dtype=torch.float16 ).to('cuda:0')
> > pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)` ` image1 = Image.fromarray(image) image = np.array(image) low_threshold = 100 high_threshold = 200 canny = cv2.Canny(image, low_threshold, high_threshold) canny = canny[:, :, None] canny = np.concatenate([canny, canny, canny], axis=2) canny_image = Image.fromarray(canny) result_img = pipe(ref_image=image1, prompt=prompt, negative_prompt="lowres, two head, bad anatomy, bad hands, text, error, missing fingers, extra digit, fewer digits, cropped, worst quality, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry", image=canny_image, num_inference_steps=20, controlnet_conditioning_scale=0.8, reference_attn=True, reference_adain=True, style_fidelity=0.5).images[0]`
> 
> I still face with the same issue. Could you please give me any advices? @okotaku

normalized the reference image into [-1,1] not [0,1]

**Comment by patrickvonplaten** [comment]
@DN6 there is still lots of interest I think. Wonder if we should circle back to a naive diffusers integration as proposed in https://github.com/huggingface/diffusers/issues/4257

**Pull request merged**